src/traces/bar/calc.js

```diff
@@ -51,15 +51,21 @@ module.exports = function calc(gd, trace) {
     var serieslen = Math.min(pos.length, size.length),
         cd = [];
 
-    // set position
+    // set position and size
     for(i = 0; i < serieslen; i++) {
+        var p = pos[i];
+        var s = size[i];
 
         // add bars with non-numeric sizes to calcdata
         // so that ensure that traces with gaps are
         // plotted in the correct order
 
-        if(isNumeric(pos[i])) {
-            cd.push({p: pos[i]});
+        if(isNumeric(p)) {
+            if(isNumeric(s)) {
+                cd.push({ p: p, s: s });
+            } else {
+                cd.push({ p: p });
+            }
```

Comment:
Looks like scatter keeps a 1:1 mapping between calcdata points and the input arrays, filling in non-numerics with [`{x: false, y: false}`](https://github.com/plotly/plotly.js/blob/master/src/traces/scatter/calc.js#L117) - that's a good point about other `arrayOK` attributes, maybe it would be easier to keep that 1:1 mapping for bars too? Should be faster too, as we're not doing repeated `.push`, can set the size of `cd` at the start.